cpu/kinetis_common/periph/spi.c

```diff
@@ -36,1543 +20,180 @@
-                   | SPI_MCR_DOZE_MASK
-                   | SPI_MCR_CLR_TXF_MASK
-                   | SPI_MCR_CLR_RXF_MASK;
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    /* lock and power on the bus */
+    mutex_lock(&locks[bus]);
+    poweron(bus);
 
-    if (KINETIS_SPI_USE_HW_CS) {
-        spi_dev->MCR |= SPI_MCR_PCSIS(1);
-    }
+    /* enable the device */
+    dev(bus)->MCR &= ~(SPI_MCR_HALT_MASK | SPI_MCR_MDIS_MASK);
 
-    spi_dev->RSER = (uint32_t)0;
+    /* configure clock and mode */
+    dev(bus)->CTAR[0] = (mode | SPI_CTAR_FMSZ(7) | spi_clk_config[clk]);
 
-    return 0;
+    return SPI_OK;
 }
 
-int spi_init_slave(spi_t dev, spi_conf_t conf, char(*cb)(char data))
+void spi_release(spi_t bus)
 {
-    SPI_Type *spi_dev;
+    /* disable, power off, and unlock the bus */
+    dev(bus)->MCR |= (SPI_MCR_HALT_MASK | SPI_MCR_MDIS_MASK);
+    poweroff(bus);
+    mutex_unlock(&locks[bus]);
+}
 
-    switch (dev) {
-#if SPI_0_EN
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    uint8_t *out_buf = (uint8_t *)out;
+    uint8_t *in_buf = (uint8_t *)in;
+    uint32_t flags = SPI_PUSHR_CTCNT_MASK;
 
-        case SPI_0:
-            spi_dev = SPI_0_DEV;
-            /* enable clocks */
-            SPI_0_CLKEN();
-            SPI_0_PCS0_PORT_CLKEN();
-            SPI_0_SCK_PORT_CLKEN();
-            SPI_0_SOUT_PORT_CLKEN();
-            SPI_0_SIN_PORT_CLKEN();
-            /* Set PORT to AF mode */
-            SPI_0_PCS0_PORT->PCR[SPI_0_PCS0_PIN] = PORT_PCR_MUX(SPI_0_PCS0_AF);
-            SPI_0_SCK_PORT->PCR[SPI_0_SCK_PIN] = PORT_PCR_MUX(SPI_0_SCK_AF);
-            SPI_0_SOUT_PORT->PCR[SPI_0_SOUT_PIN] = PORT_PCR_MUX(SPI_0_SOUT_AF);
-            SPI_0_SIN_PORT->PCR[SPI_0_SIN_PIN] = PORT_PCR_MUX(SPI_0_SIN_AF);
-            break;
-#endif /* SPI_0_EN */
+    assert(out_buf || in_buf);
 
-        default:
-            return -1;
+    /* handle chip select */
+    if (cs != SPI_CS_UNDEF) {
+         if (cs & SWCS_MASK) {
+            gpio_clear((gpio_t)cs);
+        }
+        else {
+            flags |= (uint32_t)(1 << (cs + SPI_PUSHR_PCS_SHIFT));
+        }
     }
 
-    /* set frame size, slave mode always uses CTAR0 */
-    spi_dev->CTAR[0] = SPI_CTAR_SLAVE_FMSZ(7);
-
-    /* Set clock polarity and phase. */
-    switch (conf) {
-        case SPI_CONF_FIRST_RISING:
-            spi_dev->CTAR[0] &= ~(SPI_CTAR_CPHA_MASK | SPI_CTAR_CPOL_MASK);
-            break;
-
-        case SPI_CONF_SECOND_RISING:
-            spi_dev->CTAR[0] &= ~(SPI_CTAR_CPOL_MASK);
-            spi_dev->CTAR[0] |= SPI_CTAR_CPHA_MASK;
-            break;
-
-        case SPI_CONF_FIRST_FALLING:
-            spi_dev->CTAR[0] &= ~(SPI_CTAR_CPHA_MASK);
-            spi_dev->CTAR[0] |= SPI_CTAR_CPOL_MASK;
-            break;
+    for (size_t i = 0; i < len; i++) {
+        uint8_t tmp = (out_buf) ? out_buf[i] : 0;
+        if ((i == (len - 1)) && (!cont)) {
```

Comment:
the (!cont) here is correct, we the `SPI_PUSHR_CONT_MASK` should only be cleared if `cont := false`